prepline_general/api/general.py

```diff
@@ -58,28 +59,115 @@ def is_expected_response_type(media_type, response_type):
     os.environ["UNSTRUCTURED_ALLOWED_MIMETYPES"] = DEFAULT_MIMETYPES
 
 
+def get_pdf_splits(pdf, split_size=1):
+    """
+    Given a pdf (PdfReader) with n pages, split it into pdfs each with split_size # of pages
+    Return the files as a list of BytesIO
+    """
+    split_pdfs = []
+
+    count = 0
+
+    while count < len(pdf.pages):
+        new_pdf = PdfWriter()
+        pdf_buffer = io.BytesIO()
+
+        end = count + split_size
+        for page in pdf.pages[count:end]:
+            new_pdf.add_page(page)
+
+        new_pdf.write(pdf_buffer)
+        pdf_buffer.seek(0)
+
+        split_pdfs.append(pdf_buffer)
+        count += split_size
+
+    return split_pdfs
+
+
+def partition_file_via_api(file, **kwargs):
+    """
+    Given a file-like, use partition_via_api to retrieve its elements.
+    Hardcode sending it to api.unstructured.io for now.
+    """
+    # Note (austin) - the correct thing is to call back to ourselves
+    # using the request.url. We'll need to pass the Request into pipeline_api().
+    # Same goes for any request context to pass on, i.e. the token
+    request_url = "https://api.unstructured.io/general/v0/general"
+
+    try:
+        # Note (austin) - consider using partition_multiple_via_api to cut down on traffic
+        # This would serialize the work that we just tried to split up,
+        # so we first need to parallelize multi file input
+        elements = partition_via_api(file=file, api_url=request_url, **kwargs)
+    except ValueError as e:
+        # TODO (austin) - the status code comes back in an error message
+        # Need to pull that back out and set it here
+        raise HTTPException(status_code=500, detail=f"{e}")
+
+    return elements
+
+
+def partition_pdf_splits(file, **kwargs):
+    """
+    Split up a pdf and process in parallel by sending back into the api
+    """
+    pages_per_pdf = 1
+    pdf = PdfReader(file)
+
+    # If it's small enough, just process locally
+    if len(pdf.pages) <= pages_per_pdf:
+        return partition(file=file, **kwargs)
+
+    results = []
+    pages = get_pdf_splits(pdf, split_size=pages_per_pdf)
+    partition_func = partial(partition_file_via_api, **kwargs)
+
+    with ThreadPoolExecutor() as executor:
+        for result in executor.map(partition_func, pages):
+            results.extend(result)
+
+    return results
+
+
 def pipeline_api(
     file,
     filename="",
     m_strategy=[],
     m_coordinates=[],
+    m_pdf_processing_mode=[],
     file_content_type=None,
     response_type="application/json",
 ):
     strategy = (m_strategy[0] if len(m_strategy) else "fast").lower()
-    if strategy not in ["fast", "hi_res"]:
+    strategies = ["fast", "hi_res", "auto"]
```

Comment:
good catch on "auto"!

Reply:
wonder (longterm, not this PR) if these should be defined in a way that doesn't result in loose lists of strings like this. i.e. enum or something alongside original definition that forces us to conform.